### PR TITLE
[#IOPID-351] Refactor session verification for LV

### PR DESCRIPTION
**Pull request by gquadrati**
#### Motivation and Context

In order to verify the new LV user session, we need to check the presence of redis key `KEYS-{CF}`.
To do so, we needed to update the key format, storing the login type (LV / LEGACY) 

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/types/assertionRef.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  assertionRef: AssertionRef,
+  version: t.literal(1),
+});
+
+// StoredAssertionRefV2
+export type StoredAssertionRefV2T = t.TypeOf<typeof StoredAssertionRefV2>;
+export const StoredAssertionRefV2 = t.type({
+  assertionRef: AssertionRef,
+  version: t.literal(2),
+});
```
Comment (BurnedMarshal) [inline src/types/assertionRef.ts:18]: This object will be stored one time for user (milions of records) on Redis. Maybe we should consider a more compact format to consume less memory. For example:
```typescript
export const StoredAssertionRefV1 = t.type({
  a: AssertionRef,
  v: t.literal(1),
});
```
Or alternatively  we can improve the encoder decoder to produce a JSON with that form (I prefer this option)

File: src/types/assertionRef.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  CompactLollipopDataT,
+  CompactLollipopDataT
+>(
+  "LollipopDataFromCompact",
+  LollipopData.is,
+  (val, _context) =>
+    pipe(
+      val,
+      (data) => ({ assertionRef: data.a, loginType: data.t }),
+      LollipopData.decode
```
Comment (BurnedMarshal) [inline src/types/assertionRef.ts:35]: Maybe is equivalent, but the error message could be more consistent if types composition will be needed.
```suggestion
      (data) => LollipopData.validate(data, _context)
```

File: src/services/__tests__/redisSessionStorage.test.ts [hunk 3]
```diff
@@ -1396,6 +1398,89 @@ describe("RedisSessionStorage#userHasActiveSessions", () => {
...
+          sessionToken: aValidUser.session_token,
+        })
+      )
+    );
+    mockGet.mockImplementationOnce((_, __) =>
+      Promise.resolve(JSON.stringify(aValidUser))
+    );
+
+    const result = await sessionStorage.isUserLogged(aValidUser.fiscal_code);
+    expect(result).toEqual(E.right(true));
```
Comment (BurnedMarshal) [inline src/services/__tests__/redisSessionStorage.test.ts:1435]: To verify the logic add a couple of expect like:
```
expect(mockGet).toHaveBeenNthCalledWith(1, `KEYS-${aValidUser.fiscal_code}`);
expect(mockSmembers).toBeCalledWith(aValidUser.fiscal_code);
expect(mockGet).toHaveBeenNthCalledWith(2, `SESSIONINFO-${aValidUser.session_token}`)
```

File: src/services/__tests__/redisSessionStorage.test.ts [hunk 4]
```diff
@@ -1396,6 +1398,89 @@ describe("RedisSessionStorage#userHasActiveSessions", () => {
...
+  });
+
+  it("should return false if no LollipopData was found", async () => {
+    mockGet.mockImplementationOnce((_, __) => Promise.resolve(null));
+
+    const result = await sessionStorage.isUserLogged(aValidUser.fiscal_code);
+    expect(result).toEqual(E.right(false));
+  });
+
+  it("should return false if login type is LEGACY and user has no active sessions", async () => {
```
Comment (BurnedMarshal) [inline src/services/__tests__/redisSessionStorage.test.ts:1445]: There is another scenario that should be tested. The SessionInfo has not an expire but the Session has it. We should test when the SMembers return a `SESSIONINFO` value and the next session get doesn't find a session.

File: src/services/__tests__/redisSessionStorage.test.ts [hunk 5]
```diff
@@ -1613,9 +1698,24 @@ describe("RedisSessionStorage#getLollipopAssertionRefForUser", () => {
 
     expect(mockGet).toHaveBeenCalledTimes(1);
     expect(mockGet).toBeCalledWith(`KEYS-${aValidUser.fiscal_code}`);
-    expect(E.isRight(response)).toBeTruthy();
-    if (E.isRight(response) && O.isSome(response.right))
-      expect(response.right.value).toEqual(anAssertionRef);
+
+    expect(response).toEqual(E.right(O.some(anAssertionRef)));
```
Comment (BurnedMarshal) [inline src/services/__tests__/redisSessionStorage.test.ts:1702]: kudos, much better :D

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: src/services/redisSessionStorage.ts [hunk 6]
```diff
@@ -513,6 +516,45 @@ export default class RedisSessionStorage
     )();
   }
 
+  /**
+   * Check if user id logged in, by checking the presence of LollipopData
+   *
+   * @param fiscalCode
+   * @returns true if login type is LV or a LEGACY session exists, false otherwise
+   */
+  public async isUserLogged(
```
[the remark this answers is not in the document]
Comment (gquadrati) [reply src/services/redisSessionStorage.ts]: In that case the login type would be `LEGACY`

File: src/controllers/notificationController.ts [hunk 7]
```diff
@@ -56,10 +56,7 @@ export default class NotificationController {
         (data: Notification) =>
           pipe(
             TE.tryCatch(
-              () =>
-                this.sessionStorage.userHasActiveSessions(
-                  data.message.fiscal_code
-                ),
+              () => this.sessionStorage.isUserLogged(data.message.fiscal_code),
```
[the remark this answers is not in the document]
Comment (gquadrati) [reply src/controllers/notificationController.ts]: Done in [4386d13](https://github.com/pagopa/io-backend/pull/1031/commits/4386d13b6f1a620054db37e2b4ec05d012dc6157)

File: src/types/assertionRef.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
+import * as t from "io-ts";
+import { pipe } from "fp-ts/lib/function";
+import { JsonFromString } from "io-ts-types";
+
+import { LoginType } from "../utils/fastLogin";
+
+import { AssertionRef } from "../../generated/lollipop-api/AssertionRef";
+
+// LollipopData
+export type LollipopDataT = t.TypeOf<typeof LollipopData>;
```
[the remark this answers is not in the document]
Comment (gquadrati) [reply src/types/assertionRef.ts]: I think it's more readable. Anyway, removed in [d771338](https://github.com/pagopa/io-backend/pull/1031/commits/d7713385dcdf430a89c21350858db77c2e46cbfd)

File: src/services/redisSessionStorage.ts [hunk 9]
```diff
@@ -818,7 +881,7 @@ export default class RedisSessionStorage
           this.redisClient.setEx(
             `${lollipopFingerprintPrefix}${user.fiscal_code}`,
```
[the remark this answers is not in the document]
Comment (gquadrati) [reply src/services/redisSessionStorage.ts]: Done in [d771338](https://github.com/pagopa/io-backend/pull/1031/commits/d7713385dcdf430a89c21350858db77c2e46cbfd)

File: src/services/redisSessionStorage.ts [hunk 10]
```diff
@@ -784,15 +826,36 @@ export default class RedisSessionStorage
...
+    return pipe(
+      await this.getLollipopDataForUser(fiscalCode),
+      E.map(O.map((data) => data.assertionRef))
+    );
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  // eslint-disable-next-line @typescript-eslint/member-ordering
```
[the remark this answers is not in the document]
Comment (gquadrati) [reply src/services/redisSessionStorage.ts]: Done in [d771338](https://github.com/pagopa/io-backend/pull/1031/commits/d7713385dcdf430a89c21350858db77c2e46cbfd)

**Comment by gquadrati** [comment]
Replaced by https://github.com/pagopa/io-backend/pull/1036

**Pull request closed without merging**